woocommerce/woocommerce-android
# Issue/5416 refunding fees

Closes: #5416 


Merge Instructions:
1. [Review PR in FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2226)
2. Review this PR
3. Merge FluxC PR
4. Update FluxC hash + remove "Do not merge" label
5. Merge this PR

### Description


Goal of this PR is to add support for refunding fees from the app.

I'm sorry for the size of this PR. I'm usually against PRs of this size, however, 90+% of the code is a simply copy-paste of the logic for shipping refunds. **I recommend reviewing this PR by commits to follow my train of thoughts.**

Commits 00c27d87782d018730130eb329f5362191d4b1c6 and fbb51eb9470390a3e8a9ef029ce3c4f79074faf2 are by far the largest, but they contain only a copy-paste of shipping refunds logic + renaming. They should still be reviewed, since I might have made a mistake, but it's not necessary to review the logic.

Notes:
1. The code contains even a RecyclerView (including the adapter, layouts, view logic) that lists all the FeeLines of the order. However, the visibility of this recycler view is hard coded to GONE so the code is technically never used. I wasn't sure if I should include a dead code. However, I opted for creating a copy of the behavior of Shipping refunds. If we decide to support refunding fees by row, it'll be just a matter of making the RecyclerView and Switcher on list items visible. I plan to create a project proposal which will consolidate all this code - either remove it or enable the support.
2. I've noticed a few issues with Refunds when I was working on this PR - I'll elaborate on these on the project proposal.
- "Issue Refund" button is not visible when only shipping has not been refunded yet 
- "You can refund taxes in your store admin" label is visible. However, the app allows the user to refund taxes - all shipping/fee/product taxes. It causes some unexpected side-effects in wpadmin.
- Product refund section is visible even when the order doesn't have any products - eg. it contains only fees (such as Simple Payments Order)
3. I added just a very basic unit tests - the VM is missing most of the unit tests and I plan to cover that in the project proposal too


### Testing instructions


1. Open detail of an order with a fee item available for refund
2. Tap on "Issue Refunds" button
3. Verify that the Refund Fees Section is displayed as a single row containing a disabled toggle
4. Activate the Toggle and verify that the full Refund Fees section shows up, displaying currently the subtotal, tax and total fields, also verify that the Next button becomes clickable too.
5. Hit the Next button and verify if the Issue Confirmation view show up correctly as expected, hit the Refund button and check if the Refund happened as expected

- Repeat the same steps as above on an order with multiple fee items

### Images/gif

| 1 | 2 |
| --- | --- |
| ![Screenshot_1640265239](https://user-images.githubusercontent.com/2261188/147245618-99e45069-ddbb-4e7b-82f5-ebe54d883a6b.png) | ![Screenshot_1640265240](https://user-images.githubusercontent.com/2261188/147245641-03a5542d-074f-4682-b538-f7c35f1dca08.png) |
